### PR TITLE
Removed GC blobs from channel context (DDS). The GC blobs in data store is used to generate initial GC data for DDS

### LINKED ISSUE #4943: Reduce # of GC blobs in a document by removing GC blobs from DDS's snapshot
The GC data for a DDS that is written in it's GC blob is also available in the GC blob of its parent data store.
It might be possible to generate the DDS's GC data from the data store's GC data and pass it during the data store realization.

This will greatly reduce the number of blobs in the snapshot sent to odsp.

### PR DESCRIPTION
**Pull request by agarwal-navin**
Fixes #4943.

- Removed GC blobs from channel contexts (DDS).
- The GC blobs in the data store is now used to generate the initial GC data for channel contexts.
- The corresponding change in the FluidFrameworkTestData repo - https://github.com/microsoft/FluidFrameworkTestData/pull/11

Couple of notes:
- This change introduces a small inefficiency when a new DDS is created after the data store is attached. When the summarizer tries to get its GC data for the first time, it will generate the GC data even if there was no change in the DDS since it got the attach message. This is because the initial summary of the data store will not have the data for this DDS.
  - There is a simple fix for this - Add the GC blob to the DDS's summary only when creating attach message. However, this will introduce extra GC blobs in the initial summary which won't get removed until this DDS is summarized again. The inefficiency seems worth this trade-off. We can always add this if we realize that summarizing the DDS is becoming too expensive.
- Documents before this change won't have the `getInitialGCSummaryDetails` in `IFluidChannelContext`. So, the summarizer would end up summarizing these DDSs the first time. I can handle this scenario by adding some back-compat code but I don't think its worth maintaining that code for DDSs. Please let me know if you think we need to handle this scenario.

### CODE REVIEW

File: packages/runtime/datastore/src/dataStoreRuntime.ts [hunk 1]
```diff
@@ -267,6 +272,61 @@ IFluidDataStoreChannel, IFluidDataStoreRuntime, IFluidHandleContext {
...
+            // Remove route to this data store that was added when generating GC nodes.
+            childGCData.gcNodes[childNodeId === "" ? "/" : childNodeId] = outboundRoutes.filter(
+                (route: string) => { return route !== this.absolutePath; },
+            );
+            childMap.set(childId, childGCData);
+        }
+        return childMap;
+    }
+
+    private async getChannelInitialGCDetails(channelId: string): Promise<IGarbageCollectionSummaryDetails> {
```
Comment (vladsud) [inline packages/runtime/datastore/src/dataStoreRuntime.ts:301]: please put comments above functions on what they do.

File: packages/runtime/datastore/src/remoteChannelContext.ts [hunk 2]
```diff
@@ -53,7 +53,11 @@ export class RemoteChannelContext implements IChannelContext {
      * This loads the GC details from the base snapshot of this context.
      */
     private readonly gcDetailsInInitialSummaryP = new LazyPromise<IGarbageCollectionSummaryDetails>(async () => {
-        if (await this.services.objectStorage.contains(gcBlobKey)) {
+        // back-compat 0.34.0 - Older documents may have the GC blob in the channel context's summary. Much older
+        // documents may not have GC blobs at all.
+        if (this.gcDetailsInInitialSummary !== undefined) {
```
Comment (vladsud) [inline packages/runtime/datastore/src/remoteChannelContext.ts:58]: I do not think this check is required - there is no back compat concerns between data store runtime and DDSs - they exists in the same layer.

**Comment by vladsud** [comment]
Looks reasonable to me.

File: packages/runtime/datastore/src/dataStoreRuntime.ts [hunk 3]
```diff
@@ -638,6 +685,30 @@ IFluidDataStoreChannel, IFluidDataStoreRuntime, IFluidHandleContext {
...
     }
 
+    /**
+     * Returns the GC details in initial summary for the channel with the given id. The initial summary of the data
+     * store contains the GC details of all the child channel contexts that were created before the summary was taken.
+     * We find the GC details belonging to the given channel context and return it.
+     * @param channelId - The id of the channel context that is asked for the initial GC details.
+     * @returns the requested channel's GC details in the initial summary.
+     */
+    private async getChannelInitialGCDetails(channelId: string): Promise<IGarbageCollectionSummaryDetails> {
```
Comment (vladsud) [inline packages/runtime/datastore/src/dataStoreRuntime.ts:695]: I assume we have reasonable test coverage for new logic through existing UTs, right?

**Review by vladsud — APPROVED** [review]

File: packages/runtime/datastore/src/dataStoreRuntime.ts (hunk 3 above)
Comment (agarwal-navin) [reply packages/runtime/datastore/src/dataStoreRuntime.ts:695]: Unfortunately, we do not have unit tests for data store runtime! and for channel contexts too. We should definitely add those.

Comment (agarwal-navin) [reply packages/runtime/datastore/src/dataStoreRuntime.ts:695]: Created #5057 to add testing for these layers.

**Pull request merged**